game/i18n.hh

```diff
@@ -2,26 +2,130 @@
 public:
-	TranslationEngine(const char *package) {
-		boost::locale::generator gen;
-		gen.add_messages_path(getLocaleDir().string());
-		gen.add_messages_domain(package);
-		try {
-			std::locale::global(gen(boost::locale::util::get_system_locale(true)));
-		} catch (...) {
-			std::clog << "locale/warning: Unable to detect locale, will try to fallback to en_US.UTF-8" << std::endl;
-			std::locale::global(gen("en_US.UTF-8"));
-		}
+	TranslationEngine(const char *package) : m_package(package) {
+
+		initializeAllLanguages();
+		populateLanguages(GetAllLanguages());
+
+		std::clog << "locale/debug: Checking if language is set in config." << std::endl;
+
+		auto prefValue = config["game/language"].getEnumName();
+		auto lang = getLanguageByHumanReadableName(prefValue);
+
+		setLanguage(lang);
 	};
 	static bool enabled() {
 		return true;
 	};
 
+	static ConfigItem& translationConfig() {
+		static ConfigItem& backend = config["game/language"];
+		return backend;
+	}
+
+	void initializeAllLanguages() {
+		auto path = getLocaleDir().string();
+		m_gen.add_messages_path(path);
+		m_gen.add_messages_domain(m_package);
+		m_gen.locale_cache_enabled(true);
+
+		for (auto const& language : GetAllLanguages()) {
+			if (language.first == "None") continue;
+			m_gen(language.first);
+		}
+	}
+
+	void setLanguage(const std::string& language, bool fromSettings = false) {
+		std::clog << "locale/notice: Setting language to: '" << language << "'" << std::endl;
+
+		if (fromSettings) {
+			m_currentLanguage = { getLanguageByHumanReadableName(language) , language };
+		}
+		else {
+			m_currentLanguage = { language , getLanguageByKey(language) };
+		}
+
+		try {
+			std::locale::global(m_gen(m_currentLanguage.first));
+			std::cout << "locale/notice: Current language is: '" << m_currentLanguage.second << "'" << std::endl;
+			populateLanguages(GetAllLanguages(true), true);
+		}
+		catch (...) {
+			std::clog << "locale/warning: Unable to detect locale, will try to fallback to en_US.UTF-8" << std::endl;
+			std::locale::global(m_gen("en_US.UTF-8"));
+		}
+	}
+
+	std::string getLanguageByHumanReadableName(const std::string& language) {
+		if (language == _("Auto")) {
+			return boost::locale::util::get_system_locale(true);
+		}
+
+		auto allLanguages = GetAllLanguages();	
+		auto languageKey = std::string();
+		for (auto it = allLanguages.begin(); it != allLanguages.end(); ++it) {
+			if (it->second == language) {
+				languageKey = it->first;
+				break;
+			}
+		}
+
+		return languageKey;
+	}
+
+	std::string getLanguageByKey(const std::string& languageKey) {
+		if (languageKey == "None") {
+			return boost::locale::util::get_system_locale(true);
+		}
+
+		auto allLanguages = GetAllLanguages();
+		if (allLanguages.find(languageKey) == allLanguages.end()) {
+			return allLanguages["en_US.UTF-8"];
+		}
+		return allLanguages.at(languageKey);
+	}
+
+	const std::pair<std::string, std::string> getCurrentLanguage() { return m_currentLanguage; }
```

Comment:
if you really want by value
```suggestion
	 std::pair<std::string, std::string> getCurrentLanguage() { return m_currentLanguage; }
```
or by reference:
```suggestion
	const std::pair<std::string, std::string> &getCurrentLanguage() { return m_currentLanguage; }
```

Reply:
Did it by reference